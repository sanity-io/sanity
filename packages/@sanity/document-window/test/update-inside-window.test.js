const DocumentWindow = require('../src/DocumentWindow')
const {getSnapshotFixture} = require('./fixtures')
const {getMockClient, gatherWindows, mockMutation} = require('./helpers')

const Query = DocumentWindow.Query
let docWindow = null

afterEach(() => {
  if (docWindow) {
    docWindow.removeAllListeners()
  }
})

describe('updates inside of range', () => {
  test('updates into window changes item', async () => {
    const newDocument = {
      _id: 'cdefghijklmnopqrstuvwxyz',
      name: 'New title',
      _updatedAt: '2017-06-02T12:00:00Z',
      __updated: true
    }

    const query = new Query().from(0).to(5).orderBy('_updatedAt', 'desc')
    const client = getMockClient({
      responses: [getSnapshotFixture(0, 10)],
      events: [mockMutation(newDocument)]
    })

    docWindow = new DocumentWindow({client, query})
    const windows = await gatherWindows(docWindow, 2)

    expect(windows[0]).toMatchSnapshot('01. a) window prior to update')
    expect(windows[1]).toMatchSnapshot('01. b) window after update')
    expect(docWindow).toMatchSnapshot('01. c) buffers after update')
  })

  test('updates into window with changed sort changes order', async () => {
    const newDocument = {
      _id: '2-wvutsrqponmlkjihgfedcba',
      name: 'That new product',
      _updatedAt: '2017-06-02T11:59:44Z',
      __updated: true
    }

    const query = new Query().from(12).to(17).orderBy('_updatedAt', 'desc')
    const client = getMockClient({
      responses: [getSnapshotFixture(0, 35)],
      events: [mockMutation(newDocument)]
    })

    docWindow = new DocumentWindow({client, query})
    const windows = await gatherWindows(docWindow, 2)

    expect(windows[0]).toMatchSnapshot('02. a) window prior to update')
    expect(windows[1]).toMatchSnapshot('02. b) window after update')
    expect(windows[0][2]).toBe(windows[1][1])
  })
})
