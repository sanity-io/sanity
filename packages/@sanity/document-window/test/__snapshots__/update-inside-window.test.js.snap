// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`01. a) window prior to update 1`] = `
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 2",
    "productIndex": 2,
  },
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
  Object {
    "_id": "efghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 4",
    "productIndex": 4,
  },
]
`;

exports[`01. b) window after update 1`] = `
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  
  ----[UPDATED]----
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "New title",
  }
  ---[/UPDATED]----

,
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
  Object {
    "_id": "efghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 4",
    "productIndex": 4,
  },
]
`;

exports[`01. c) buffers after update 1`] = `
[=== Pre ===]
Array []
[=== /Pre ===]

[=== Window ===]
Array [
  Object {
    "_id": "abcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 0",
    "productIndex": 0,
  },
  Object {
    "_id": "bcdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 1",
    "productIndex": 1,
  },
  
  ----[UPDATED]----
  Object {
    "_id": "cdefghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "New title",
  }
  ---[/UPDATED]----

,
  Object {
    "_id": "defghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 3",
    "productIndex": 3,
  },
  Object {
    "_id": "efghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 4",
    "productIndex": 4,
  },
]
[=== /Window ===]

[=== Post ===]
Array [
  Object {
    "_id": "fghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 5",
    "productIndex": 5,
  },
  Object {
    "_id": "ghijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 6",
    "productIndex": 6,
  },
  Object {
    "_id": "hijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 7",
    "productIndex": 7,
  },
  Object {
    "_id": "ijklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 8",
    "productIndex": 8,
  },
  Object {
    "_id": "jklmnopqrstuvwxyz",
    "_updatedAt": "2017-06-02T12:00:00Z",
    "name": "Product 9",
    "productIndex": 9,
  },
]
[=== /Post ===]

`;

exports[`02. a) window prior to update 1`] = `
Array [
  Object {
    "_id": "2-xwvutsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:51Z",
    "name": "Product 12",
    "productIndex": 12,
  },
  Object {
    "_id": "2-wvutsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:48Z",
    "name": "Product 13",
    "productIndex": 13,
  },
  Object {
    "_id": "2-vutsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:45Z",
    "name": "Product 14",
    "productIndex": 14,
  },
  Object {
    "_id": "2-utsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:42Z",
    "name": "Product 15",
    "productIndex": 15,
  },
  Object {
    "_id": "2-tsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:39Z",
    "name": "Product 16",
    "productIndex": 16,
  },
]
`;

exports[`02. b) window after update 1`] = `
Array [
  Object {
    "_id": "2-xwvutsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:51Z",
    "name": "Product 12",
    "productIndex": 12,
  },
  Object {
    "_id": "2-vutsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:45Z",
    "name": "Product 14",
    "productIndex": 14,
  },
  
  ----[UPDATED]----
  Object {
    "_id": "2-wvutsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:44Z",
    "name": "That new product",
  }
  ---[/UPDATED]----

,
  Object {
    "_id": "2-utsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:42Z",
    "name": "Product 15",
    "productIndex": 15,
  },
  Object {
    "_id": "2-tsrqponmlkjihgfedcba",
    "_updatedAt": "2017-06-02T11:59:39Z",
    "name": "Product 16",
    "productIndex": 16,
  },
]
`;
